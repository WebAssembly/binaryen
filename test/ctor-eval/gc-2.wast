(module
  (type $struct (struct_subtype (field i32) data))

  (import "import" "import" (func $import (param anyref)))

  ;; This struct is created in an immutable global, but it has the wrong type.
  ;; We will create a new defining global for it that has the proper type, and
  ;; read from it here. (This is necessary as when the global is used elsewhere
  ;; we want to get the right type from the global.get.)
  (global $global1 (ref any)
    (struct.new $struct
      (i32.const 1337)
    )
  )

  ;; Test reordering of globals. This global will be written a value that is
  ;; actually defined after it. To handle that, we must create it earlier than
  ;; this global.
  (global $global2 (mut (ref null $struct))
    (ref.null $struct)
  )

  ;; This global is perfect to be a defining global (immutable, right type), but
  ;; because of an earlier use, we will end up defining it earlier on, and
  ;; reading it here.
  (global $global3 (ref $struct)
    (struct.new $struct
      (i32.const 9999)
    )
  )

  (func "test1"
    (global.set $global2
      (global.get $global3)
    )
  )

  (func "keepalive" (result i32)
    (select
      (struct.get $struct 0
        (ref.cast_static $struct
          (global.get $global1)
        )
      )
      (struct.get $struct 0
        (global.get $global2)
      )
      (struct.get $struct 0
        (global.get $global3)
      )
    )
  )
)

