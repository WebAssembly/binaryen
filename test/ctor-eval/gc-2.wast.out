(module
 (type $struct (struct (field i32)))
 (type $none_=>_i32 (func (result i32)))
 (global $ctor-eval$global (ref $struct) (struct.new $struct
  (i32.const 1337)
 ))
 (global $global1 (ref any) (global.get $ctor-eval$global))
 (global $ctor-eval$global_0 (ref $struct) (struct.new $struct
  (i32.const 9999)
 ))
 (global $global2 (mut (ref null $struct)) (global.get $ctor-eval$global_0))
 (global $global3 (ref $struct) (global.get $ctor-eval$global_0))
 (export "keepalive" (func $1))
 (func $1 (type $none_=>_i32) (result i32)
  (select
   (struct.get $struct 0
    (ref.cast $struct
     (global.get $global1)
    )
   )
   (struct.get $struct 0
    (global.get $global2)
   )
   (struct.get $struct 0
    (global.get $global3)
   )
  )
 )
)
